### PR TITLE
CNV-20183- new content temporary token VNC

### PR DESCRIPTION
**Pull request by sabrinajess**
Version(s): 4.14+

Issue: [CNV-20183](https://issues.redhat.com//browse/CNV-20183)

Link to docs preview: https://64453--docspreview.netlify.app/openshift-enterprise/latest/virt/virtual_machines/virt-accessing-vm-consoles#virt-temporary-token-VNC_vnc-console

QE review:
- [x] QE has approved this change.

Additional information:

### CODE REVIEW

File: modules/virt-temporary-token-VNC.adoc [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+:_content-type: PROCEDURE
+[id="virt-temporary-token-VNC_{context}"]
+= Generating a temporary token for VNC console
+
+Generate a temporary token to access the VNC of a virtual machine (VM).
+
+.Prerequisites
+
+* A running Windows virtual machine
+* ?
```
Comment (akrejcir) [inline modules/virt-temporary-token-VNC.adoc:14]: This feature is not related in to Windows. It works for all VMs.

File: modules/virt-temporary-token-VNC.adoc [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+$ export VNC_TOKEN="<token>"
+----
+
+.  Log in to the cluster:
++
+[source,terminal]
+----
+$ oc login --token ${VNC_TOKEN}
+----
+
```
Comment (akrejcir) [inline modules/virt-temporary-token-VNC.adoc:58]: I've used the `oc login` in my demo showing this feature, because it was the simplest way to test that it works.

I expect that the token access will be used primarily by developers, that will use the token it their application. They may use `virtctl vnc` just for testing that it works.

**Review by akrejcir — COMMENTED** [review]
I will send you a google doc with a description of another way to use the token, that will probably be more useful to developers.

File: modules/virt-temporary-token-VNC.adoc [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+{ "token": "eyJhb..." }
+----
+
+. Use the token provided in the output to create a variable:
++
+[source,terminal]
+----
+$ export VNC_TOKEN="<token>"
+----
+
```
[the remark this answers is not in the document]
Comment (akrejcir) [reply modules/virt-temporary-token-VNC.adoc:51]: No.

File: modules/virt-temporary-token-VNC.adoc (hunk 2 above)
Comment (akrejcir) [reply modules/virt-temporary-token-VNC.adoc:58]: I mean that `oc login` changes the global login credentials that are used by subsequent `oc` and `virtctl` commands. User may not want to change the global credentials to a token that has only access to the VNC of one VM.
I've done it in the demo, because it was simple, but I think it is not the main use case for this feature.

> * Update the step "`Log in to the cluster`" to something similar to "`Log in to the cluster to access the VNC console of the VM`" using `oc login`.

I leave this change up to you, I don't have any opinion about which text is better.

> * Update the "Use virtctl to access the VNC console..." step to "Use virtctl to test the VNC console..." and place it in under Verification instead of Procedure.

This change sounds good, because using `virtctl vnc` with the token is not the main use case.

**Comment by sabrinajess** [comment]
@akrejcir Please review again.

File: modules/virt-temporary-token-VNC.adoc [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+* A running virtual machine with {VirtProductName} 4.14 or later and xref:../../virt/about-virt/virt-architecture#virt-about-ssp-operator_virt-architecture[`ssp-operator`] 4.14 or later
+
+.Procedure
+
+. Enable the feature gate in the HyperConverged (`HCO`) custom resource (CR):
++
+[source,yaml]
+----
+$ oc patch hyperconverged kubevirt-hyperconverged -n {CNVNamespace} \
+ --type json -p '[{"op": "replace", "path": \
```
Comment (akrejcir) [inline modules/virt-temporary-token-VNC.adoc:22]: The `\` at the end of lines should not be there. The whole patch is enclosed in single quotes `'`, so the `\` would be present in the final command, and that would make it incorrect.

File: modules/virt-temporary-token-VNC.adoc [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+----
+
+.  Log in to the cluster:
++
+[source,terminal]
+----
+$ oc login --token ${VNC_TOKEN}
+----
+
+You can now use the token to access the VNC console of a VM.
```
Comment (akrejcir) [inline modules/virt-temporary-token-VNC.adoc:58]: I would move these lines to the `Verification` part below. `oc login` is only needed if the token is used by `virtctl vnc`.

File: modules/virt-temporary-token-VNC.adoc (hunk 4 above)
Comment (sabrinajess) [reply modules/virt-temporary-token-VNC.adoc:22]: @akrejcir Would the following work?

```
$ oc patch hyperconverged kubevirt-hyperconverged -n {CNVNamespace} \
 --type json -p '[{"op": "replace", "path": ' \
 '"/spec/featureGates/deployVmConsoleProxy", "value": true}]'
```

**Comment by sabrinajess** [comment]
Personal note:
edit 
[source,yaml] 
to be [source,terminal,subs="attributes+"]

File: modules/virt-temporary-token-VNC.adoc (hunk 4 above)
Comment (akrejcir) [reply modules/virt-temporary-token-VNC.adoc:22]: No, because this would be interpreted as:
```bash
oc patch ...  -p '[{"op": "replace", "path": '  '"/spec/featureGates/deployVmConsoleProxy", "value": true}]'
```

The patch argument `-p` is broken into 2 parts separated by space:
```
' [{"op": "replace", "path": '
' "/spec/featureGates/deployVmConsoleProxy", "value": true}] '
```
but only the first part is passed as `-p` argument. The other part will probably cause an error.

Comment (sabrinajess) [reply modules/virt-temporary-token-VNC.adoc:22]: Got it -thanks!

File: modules/virt-temporary-token-VNC.adoc [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+----
+
+. Generate a token:
++
+[source,terminal]
+----
+$ curl --header "Authorization: Bearer ${TOKEN}" \
+     "https://api.<cluster_fqdn>/apis/token.kubevirt.io/v1alpha1/namespaces/<namespace>/virtualmachines/<vm_name>/vnc?duration=<duration>" <1>
+----
+<1> Duration can be in hours and/or minutes. Example: `5h30m`
```
Comment (tmalove) [inline modules/virt-temporary-token-VNC.adoc:32]:
```suggestion
<1> Duration can be in hours and minutes. Example: `5h30m`
```
See SSG: https://www.ibm.com/docs/en/ibm-style?topic=punctuation-slashes

**Review by tmalove — APPROVED** [review]

File: modules/virt-temporary-token-VNC.adoc (hunk 6 above)
Comment (geetikakay) [inline modules/virt-temporary-token-VNC.adoc:32]: should we mention what is expected when duration is set.

Comment (sabrinajess) [reply modules/virt-temporary-token-VNC.adoc:32]: @geetikakay What is expected?

Comment (geetikakay) [reply modules/virt-temporary-token-VNC.adoc:32]: duration suggest for how long the token is valid and can be used to access vnc using virtctl but I'll let @akrejcir to advice if we need or present document is fine

**Comment by sabrinajess** [comment]
@geetikakay Please review updates.

File: modules/virt-temporary-token-VNC.adoc [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+----
+$ oc patch hyperconverged kubevirt-hyperconverged -n {CNVNamespace} --type json -p '[{"op": "replace", "path": "/spec/featureGates/deployVmConsoleProxy", "value": true}]'
+# ...
+----
+
+. Generate a token:
++
+[source,terminal]
+----
+$ curl --header "Authorization: Bearer ${TOKEN}" \
```
Comment (geetikakay) [inline modules/virt-temporary-token-VNC.adoc:29]: @akrejcir Hi, do you think we should add same steps for sa creation, rolebindings that we have in demo video

File: modules/virt-temporary-token-VNC.adoc [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+     "https://api.<cluster_fqdn>/apis/token.kubevirt.io/v1alpha1/namespaces/<namespace>/virtualmachines/<vm_name>/vnc?duration=<duration>" <1>
```
[the remark this answers is not in the document]
Comment (akrejcir) [reply modules/virt-temporary-token-VNC.adoc:30]: Yes. 
@sabrinajess , we've found out during testing that minimum valid duration is 10 minutes. Shorter time will fail with an error. This is a limitation in the kubernetes, so we cannot fix it in CNV.

**Comment by kalexand-rh** [comment]
I've opened the `enterprise-4.15` branch. Please review your PR and update the applicable versions to also cover 4.15 if you expect that the changes in this PR will be valid in that branch.

**Comment by geetikakay** [comment]
lgtm !!

**Review by abhatt-rh — COMMENTED** [review]
Hi @sabrinajess 
I noticed a few nits during my merge review; added some comments for your consideration.
Also, can you please confirm if this is QE approved. Thanks!

File: modules/virt-temporary-token-VNC.adoc [hunk 9]
```diff
@@ -0,0 +1,69 @@
+// Module included in the following assemblies:
+//
+// * virt/virtual_machines/virt-accessing-vm-consoles.adoc
+
+:_content-type: PROCEDURE
+[id="virt-temporary-token-VNC_{context}"]
+= Generating a temporary token for the VNC console
+
+Generate a temporary authentication bearer token for the Kubernetes API to access the VNC of a virtual machine (VM).
```
[the remark this answers is not in the document]
Comment (sabrinajess) [reply modules/virt-temporary-token-VNC.adoc:9]: @abhatt-rh This is presently one of the exceptions in virt for spelling it out first.

**Comment by sabrinajess** [comment]
@abhatt-rh Fixed 👍

File: modules/virt-temporary-token-VNC.adoc [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+
+. Enable the feature gate in the HyperConverged (`HCO`) custom resource (CR):
++
+[source,terminal,subs="attributes+"]
+----
+$ oc patch hyperconverged kubevirt-hyperconverged -n {CNVNamespace} --type json -p '[{"op": "replace", "path": "/spec/featureGates/deployVmConsoleProxy", "value": true}]'
+# ...
+----
+
+. Generate a token:
```
Comment (abhatt-rh) [inline modules/virt-temporary-token-VNC.adoc:30]: Sorry I missed this one
```suggestion
. Generate a token by running the following command:
```

Comment (sabrinajess) [reply modules/virt-temporary-token-VNC.adoc]: Done! :)
